### PR TITLE
[PGNCCL] Record device index for GPU guarding during NCCLComm method calls

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141511
* #141510
* #141509
* __->__ #141270
* #141374

### Motivation
`ncclCommInitRank` needs GPU guard (documented in NCCL).

`ncclCommAbort`, `ncclCommFinalize` and `ncclCommDestroy` may also need GPU guard (undocumented in NCCL); otherwise, extra CUDA context may be created (or worse, hang); both effects have been seen before in our tests.

### Solution
This PR records a device index during `NCCLComm` object creation, so that we can add GPU guard in `NCCLComm`'s method calling which direct to the above NCCL APIs.

### Note
This is not a bug fix. Just a safety improvement.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/NCCLUtils.cpp [hunk 1]
```diff
@@ -65,6 +65,7 @@ std::shared_ptr<NCCLComm> NCCLComm::split(
   LOG(INFO) << "Rank " << source->rank_ << ": split from parent comm "
             << source->repr() << " with color_id " << color_id << " and rank "
             << rank;
+  at::cuda::OptionalCUDAGuard gpuGuard(source->deviceIndex_);
```
Comment (eqy) [inline torch/csrc/distributed/c10d/NCCLUtils.cpp:68]: Will this fail if called with a device index of `-1`? So we would never init a comm with an uninitialized device index, correct?

**Review by eqy — APPROVED** [review]

File: torch/csrc/distributed/c10d/NCCLUtils.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/NCCLUtils.cpp:68]: Yes, all of our created comms should have recorded a device index. If this line can fail by itself, then it is great (better than silent pass-through).

**Comment by kwen2501** [comment]
@pytorbot merge -f "CI was previously green; rebased accidentally"

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1330,18 +1330,9 @@ void ProcessGroupNCCL::abortCommsFromMap(
   for (auto& it : ncclCommsMap) {
     auto& devName = it.first;
     auto& ncclComm = it.second;
-    at::cuda::OptionalCUDAGuard gpuGuard;
-    at::DeviceIndex deviceIndex = getIndexFromDeviceKey(devName);
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1334]: Shall we also remove `getIndexFromDeviceKey`?

Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1334]: Oh do you mean this function has no use now? If so, yeah, we should remove it

**Pull request closed without merging**